flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java

```diff
@@ -0,0 +1,225 @@
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.checkpoint.CheckpointOptions;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.state.CheckpointStorageWorkerView;
+import org.apache.flink.runtime.state.CheckpointStreamFactory;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequest.completeInput;
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequest.completeOutput;
+import static org.apache.flink.runtime.checkpoint.channel.ChannelStateWriteRequest.write;
+
+/**
+ * {@link ChannelStateWriter} implemented using
+ * {@link CheckpointStreamFactory.CheckpointStateOutputStream CheckpointStateOutputStreams}. Internally, it has
+ * <ul>
+ * <li>one stream per checkpoint; having multiple streams would mean more files written and more connections opened
+ * (and more latency on restore)</li>
+ * <li>one thread; having multiple threads means more connections, couples with the implementation and increases complexity</li>
+ * </ul>
+ */
+@Internal
+@ThreadSafe
+public class ChannelStateWriterImpl implements ChannelStateWriter {
+
+	private static final Logger LOG = LoggerFactory.getLogger(ChannelStateWriterImpl.class);
+	private static final int DEFAULT_HANDOVER_CAPACITY = 10;
+	private static final int DEFAULT_MAX_CHECKPOINTS = 5; // currently, only single in-flight checkpoint is supported
+
+	private class ProcessRequestsLoop implements Runnable {
+		private final ChannelStateWriteRequestProcessor requestProcessor;
+
+		private ProcessRequestsLoop(int maxCheckpoints, CheckpointStorageWorkerView streamFactoryFactory, ChannelStateSerializer serializer) {
+			this.requestProcessor = new ChannelStateWriteRequestProcessor(maxCheckpoints, streamFactoryFactory, serializer);
+		}
+
+		@Override
+		public void run() {
+			try {
+				loop();
+			} catch (Exception ex) {
+				thrown = ex;
+			} finally {
+				handover.clear();
+				requestProcessor.cleanup(thrown != null ? thrown : new RuntimeException("loop terminated, isRunning: " + isRunning));
+			}
+			LOG.debug("loop terminated");
+		}
+
+		private void loop() throws Exception {
+			while (isRunning || !handover.isEmpty()) {
+				try {
+					requestProcessor.processRequest(handover.take());
+				} catch (InterruptedException e) {
+					if (isRunning || !handover.isEmpty()) {
+						LOG.info("interrupted while waiting for an item (continue waiting)", e);
+					} else {
+						Thread.currentThread().interrupt();
+						return;
+					}
+				}
+			}
+		}
+	}
+
+	private final Thread asyncWriter;
+	private final BlockingQueue<ChannelStateWriteRequest> handover;
```

Comment:
I am wondering  it might bring potential problems to use `BlockingQueue` with bounded size. The IO operations might be stuck sometimes in bad scenarios, then the request in the queue can not be consumed in time. Therefore it would block all the operations of `#addInputData`, especially the `#addInputData` might be triggered by multiple netty threads, to impact the network throughput. How about using unbounded queue if the memory overhead is not obvious?

Reply:
Interesting question. The usual arguments for bounded queues like protecting against memory exhaustion don't work that well here. The reasons are that the queue essentially holds references to larger memory regions and atm we can only have a single in-flight checkpoint.
However,
1. we may have multiple references to the same region (when we remove the single-checkpoint limit and don't have incremental checkpointing yet or it doesn't help here)
1. these are different memory regions (heap and likely non-heap)

With an unbounded queue, the task could run out of memory and it would be much harder to figure it out. 
So I'd stick with bounded queue but lift its default size.

WDYT?

Reply:
Thanks for sharing your thoughts. Actually I only thought of the benefits of resource limitation for bounded queue before.

Do you concerning that the network buffer to be exhausted or run out of memory by unbounded queue I f I understand correctly? If so, I guess this hurt might be less worse than blocking the netty thread for two reasons:

- If one netty thread is blocked, it might block multiple input channels receiving following data, then it might even delay the barrier alignment, based on the truth that one netty thread wouold serve for multiple input channels data transport.

- If one netty thread is blocked, it does not alway mean that the left buffer memory can be better used by other channels. For exclusive buffers per-channel, it can only be used by current blocked netty thread for receiving data. For floating buffers among all the channels, if it was already requested away by the current blocked netty thread, it also can not returned back to be used by other active netty threads. It can only avoid requesting more floating buffers by current blocked netty thread, but it is not determined behavior.

So if there are no concerns for additional memory overhead caused by unbounded queue and it is also easy to adjust it, then i prefer to this option. Otherwise we can lift the default size for bounded size and further convert to unbounded queue if finding serious problems in production future. But I am not quite sure whether the blocking impact can be easily found in practice.

Reply:
If it is easier for using bounded atm, i am also fine with setting enough capacity for the queue, then we can assume that it would never block netty thread with this large capacity.